File: Solution-Python/Combination_39.py
class Combiantion_39:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        self.helper(sorted(candidates), res, [], 0, target)
        return res
    
    def helper(self, candidates: List[int], res: List[List[int]], tmp: List[int], start: int, target: int):
        if target == 0:
            res.append(copy.deepcopy(tmp))
        else:
            for i in range(start, len(candidates)):
                if candidates[i] > target:
                    return
                tmp.append(candidates[i])
                self.helper(candidates, res, tmp, i, target - candidates[i])
                tmp.pop()

File: Solution-Go/Combination_39.go
func combinationSum(candidates []int, target int) [][]int {
    res := [][]int{}
    sort.Ints(candidates)
    helper(candidates, &res, []int{}, 0, target)
    return res
}

func helper(candidates []int, res *[][]int, tmp []int, start int, target int) {
    if target == 0 {
        b := make([]int, len(tmp))
        copy(b, tmp)
        *res = append(*res, b)
    } else {
        for i := start; i < len(candidates) && candidates[i] <= target; i++ {
            tmp = append(tmp, candidates[i])
            helper(candidates, res, tmp, i, target - candidates[i])
            tmp = tmp[:len(tmp) - 1]
        }
    }
}